### PR TITLE
Allow coupon in first cart step

### PR DESCRIPTION
**Pull request by Alban-io**

### CODE REVIEW

File: core/lib/Thelia/Exception/UnmatchableConditionException.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * @package Thelia\Exception
+ * @author Baixas Alban <abaixas@openstudio.fr>
+ */
+class UnmatchableConditionException extends \RuntimeException
+{
+    /**
+     * UnmatchableConditionException thrown when a condition can't be check (if customer is not connected)
+     */
+    public function __construct()
```
Comment (roadster31) [inline core/lib/Thelia/Exception/UnmatchableConditionException.php]: Such a constructor is useless.

File: local/modules/Front/Controller/CouponController.php [hunk 2]
```diff
@@ -123,6 +123,12 @@ public function consumeAction()
                 ["%message" => $e->getMessage()],
                 Front::MESSAGE_DOMAIN
             );
+        } catch (UnmatchableConditionException $e) {
+            $message = $this->getTranslator()->trans(
+                'You must <a href="%url">log in</a> to apply this reduction.',
+                ['%url' => $this->retrieveUrlFromRouteId('customer.login.view')],
```
Comment (roadster31) [inline local/modules/Front/Controller/CouponController.php]: "You should sign in or register to use this coupon"

File: core/lib/Thelia/Exception/UnmatchableConditionException.php (hunk 1 above)
Comment (Alban-io) [inline core/lib/Thelia/Exception/UnmatchableConditionException.php]: :white_check_mark:

File: core/lib/Thelia/Coupon/CouponManager.php [hunk 3]
```diff
@@ -218,8 +219,16 @@ protected function sortCoupons(array $coupons)
...
-                $couponsKept[] = $coupon;
+
+            try {
+
+                if ($coupon->isMatching()) {
+                    $couponsKept[] = $coupon;
+                }
+
+            } catch (UnmatchableConditionException $e) {
+                // don't kept unmatchable coupon
```
Comment (roadster31) [inline core/lib/Thelia/Coupon/CouponManager.php]: "ignore unmatchable coupon"

**Comment by roadster31** [comment]
Nice PR !

**Comment by bibich** [comment]
Thanks

**Pull request merged**